omp/reorder/rcm_kernels.cpp

```diff
@@ -0,0 +1,907 @@
+                const auto row_end = row_ptrs[written + 1];
+                const auto write_level = level + 1;
+                for (auto neighbour_i = row_start; neighbour_i < row_end;
+                     ++neighbour_i) {
+                    const auto neighbour = col_idxs[neighbour_i];
+
+                    // Will not be written by multiple threads, but can be read
+                    // while written. This is only necessary to guarantee the
+                    // abscence of reads-while-writes.
+                    IndexType neighbour_level;
+#pragma omp atomic read
+                    neighbour_level = levels[neighbour];
+
+                    // ::max() is not a valid value here (as ther cannot be
+                    // ::max() nodes), therefore can be used for signaling.
+                    if (neighbour_level == write_level) {
+// Protect against writing the same node as neighbour of different nodes in the
+// previous level by the same thread.
+#pragma omp atomic write
+                        levels[neighbour] =
+                            std::numeric_limits<IndexType>::max();
+
+                        valid_neighbours.push_back(neighbour);
+                    }
+                }
+
+                // Sort neighbours.
+                const auto size = valid_neighbours.size();
+                sort_by_degree(&valid_neighbours[0], size, degrees);
+
+                // Write the processed neighbours.
+                const auto base_write_offset =
+                    base_offset + next_level_idx + write_offset;
+
+                // Memcpy would be nice, but we need element-wise atomicity.
+                for (auto i = 0; i < size; ++i) {
+#pragma omp atomic write
+                    perm[base_write_offset + i] = valid_neighbours[i];
+                }
+
+                write_offset += size;
+                valid_neighbours.clear();
+            }
+        }
+    }
+}
+
+
+/**
+ * Processes all isolated nodes, returning their count.
+ */
+template <typename IndexType>
+IndexType handle_isolated_nodes(std::shared_ptr<const OmpExecutor> exec,
+                                const IndexType *const row_ptrs,
+                                const IndexType *const col_idxs,
+                                const IndexType *const degrees,
+                                IndexType *const perm, size_type num_vertices,
+                                vector<uint8> &previous_component)
+{
+    struct IsolatedNodes {
+        // Using a gko::vector here makes clang and gcc segfault,
+        // if we then use 'omp_priv = omp_orig'.
+        std::vector<IndexType> nodes;
+        IsolatedNodes &operator+=(const IsolatedNodes &rhs)
+        {
+            nodes.reserve(nodes.size() + rhs.nodes.size());
+            nodes.insert(nodes.end(), rhs.nodes.begin(), rhs.nodes.end());
+            return *this;
+        }
+    };
+
+#pragma omp declare reduction(FindIsolated : IsolatedNodes : omp_out += omp_in)
```

Comment:
Alternative: Unless we go to really large numbers of cores, we could do this semi-efficiently by hand with a critical section after the loop. This would also get rid of any potential segfaults with custom allocators (which sounds like OpenMP has issues with reductions on complex types)

Reply:
Just to be sure: The problem here is, that we use a `std::vector`, but would like to track the memory usage?

Reply:
That and the fact that OpenMP seems to have issues with reductions on complex types, so I would prefer if we could avoid them altogether, especially since they require such a complex framework to function.

Reply:
Just pushed a change rewriting this as manual reduction. Is this what you had in mind?